src/Substance.hs

```diff
@@ -4,535 +4,450 @@
+
+checkFuncInEnv :: VarEnv -> Func -> Operation -> (String, Maybe T)
+checkFuncInEnv varEnv (Func f args) (Operation name yls kls tls t) =
+               let errAndTypesLs = map (checkExpression varEnv) args
+                   errls         = map (\(err1,t1) -> err1) errAndTypesLs
+                   err           = foldl (\err1 err2 -> err1 ++ err2) "" errls
+                   argTypes      = map (\(err1,t1) -> t1) errAndTypesLs
+               in if (foldl (\b at1 -> b && isJust at1) True argTypes)
+                  then let argTypes2 = map (\a -> KT (fromJust a)) argTypes
+                           tls2      = map (\a -> KT a) tls
+                           sigma     = subst varEnv M.empty argTypes2 tls2
+                       in if (sigma == M.empty)
+                          then (err, Just t) -- err should be empty str
+                          else (err, Just (applySubst sigma t)) -- err should be empty str
+                  else (err, Nothing)
+
+checkVarConsInEnv  :: VarEnv -> Func -> VarConstructor -> (String, Maybe T)
+checkVarConsInEnv varEnv (Func f args) (VarConstructor name yls kls tls t) =
+                  let errAndTypesLs = map (checkExpression varEnv) args
+                      errls         = map (\(err1,t1) -> err1) errAndTypesLs
+                      err           = foldl (\err1 err2 -> err1 ++ err2) "" errls
+                      argTypes      = map (\(err1,t1) -> t1) errAndTypesLs
+                  in if (foldl (\b at1 -> b && isJust at1) True argTypes)
+                     then let argTypes2 = map (\a -> KT (fromJust a)) argTypes
+                              tls2      = map (\a -> KT a) tls
+                              sigma     = subst varEnv M.empty argTypes2 tls2
+                           in if (sigma == M.empty)
+                              then (err, Just t) -- err should be empty str
+                              else (err, Just (applySubst sigma t)) -- err should be empty str
+                     else (err, Nothing)
+
+applySubst :: M.Map Y Arg -> T -> T
+applySubst sigma (TTypeVar vt) = 
+           case sigma M.! (TypeVarY vt) of
+           AVar v -> error "Type var being mapped to variable in subst sigma, error in the TypeChecker!"
+           AT t   -> t
+applySubst sigma (TConstr (ConstructorInvoker t args pos)) =
+           let argsSub = map (applySubstHelper sigma) args
+           in TConstr (ConstructorInvoker t argsSub pos)
+
+applySubstHelper :: M.Map Y Arg -> Arg -> Arg
+applySubstHelper sigma (AVar v) = case sigma M.! (VarY v) of
+                                  res@(AVar v2) -> res
+                                  AT t -> error "Var being mapped to a type in subst sigma, error in the TypeChecker!"
+applySubstHelper sigma (AT t) = AT (applySubst sigma t)
+
+subst :: VarEnv -> M.Map Y Arg -> [K] -> [K] -> M.Map Y Arg
+subst varEnv sigma argTypes formalTypes = let types = zip argTypes formalTypes
```

Comment:
what is a "formal type"?

Reply:
would be super helpful to have just one or two concrete examples of a substitution as comments here
(e.g. what is unified? what doesn't match?)